src/app/clusters/door-lock-server/door-lock-server.cpp

```diff
@@ -140,30 +143,115 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
                                             const chip::app::ConcreteCommandPath & commandPath,
                                             const chip::app::Clusters::DoorLock::Commands::LockDoor::DecodableType & commandData)
 {
-    emberAfDoorLockClusterPrintln("Received Lock Door command (not implemented)");
+    emberAfDoorLockClusterPrintln("Received Lock Door command");
+    bool success = false;
 
-    // TODO: Implement door locking by calling emberAfPluginDoorLockOnDoorLockCommand
+    chip::EndpointId endpoint = commandPath.mEndpointId;
 
-    emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
-    return true;
+    bool require_pin = false;
+    Attributes::RequirePINforRemoteOperation::Get(endpoint, &require_pin);
+
+    if(require_pin)
+    {
+        if(!commandData.pinCode.HasValue())
+        {
+            success = false; // Just to be explicit. success == false at this point anywyay
+        }
+        else
+        {
+            const char *cmd_pin = reinterpret_cast<const char*>(commandData.pinCode.Value().data());
+            char *mPin = DoorLockServer::Instance().mPin;
+            if(strncmp(mPin, cmd_pin, 4) == 0)
+            {
+                success = emberAfPluginDoorLockOnDoorLockCommand(endpoint, cmd_pin);
+            }
+            else
+            {
+                success = false;
+            }
+        }
+    }
+    else
+    {
+        const char *cmd_pin = commandData.pinCode.HasValue() ?
+            reinterpret_cast<const char*>(commandData.pinCode.Value().data()) :
+            NULL;
+        success = emberAfPluginDoorLockOnDoorLockCommand(endpoint, cmd_pin);
+    }
+
+    emberAfSendImmediateDefaultResponse(success ? EMBER_ZCL_STATUS_SUCCESS : EMBER_ZCL_STATUS_FAILURE);
+
+    // Should be called after reponse sent (indicating command complete) since
+    // the attribute Set call in SetLockState will error out if the command is
+    // still pending
+    if(success)
+        return DoorLockServer::Instance().SetLockState(endpoint, DlLockState::kLocked) == EMBER_ZCL_STATUS_SUCCESS;
+    return success;
 }
 
 bool emberAfDoorLockClusterUnlockDoorCallback(
     chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
     const chip::app::Clusters::DoorLock::Commands::UnlockDoor::DecodableType & commandData)
 {
-    emberAfDoorLockClusterPrintln("Received Unlock Door command (not implemented)");
+    emberAfDoorLockClusterPrintln("Received Unlock Door command");
+    bool success = false;
 
-    // TODO: Implement door unlocking by calling emberAfPluginDoorLockOnDoorUnlockCommand
-    emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
-    return true;
+    chip::EndpointId endpoint = commandPath.mEndpointId;
+
+    bool require_pin = false;
+    Attributes::RequirePINforRemoteOperation::Get(endpoint, &require_pin);
+
+    if(require_pin)
```

Comment:
Similar comments throught this function as for LockDoor.

Reply:
Fixed